File: apps/cli/src/cmd/verify.ts
import { log } from "utils";
import pc from "picocolors";
import { exit } from "node-utils";
import { clearProgressMessage, writeProgress } from '../lib/terminal-utils';
import { loadDatabase, IBaseCommandOptions, ICommandContext, resolveKeyPath } from "../lib/init-cmd";
import { formatBytes } from "../lib/format";
import { verify, verifyDatabaseFiles, IDatabaseFileVerifyResult } from "api";
import { IStorageDescriptor } from "storage";
import { getS3Config } from "../lib/config";

export interface IVerifyCommandOptions extends IBaseCommandOptions {
    //
    // Force full verification (bypass cached hash optimization).
    //
    full?: boolean;

    //
    // Path to a specific file or directory to verify (instead of entire database).
    //
    path?: string;
}

//
// Command that verifies the integrity of the Photosphere media file database.
//
export async function verifyCommand(context: ICommandContext, options: IVerifyCommandOptions): Promise<void> {
    const { uuidGenerator, timestampProvider, sessionId } = context;
    const { metadataStorage, assetStorage, databaseDir, metadataCollection } = await loadDatabase(options.db, options, true, uuidGenerator, timestampProvider, sessionId);


    // Create storage descriptor for passing to workers
    const resolvedKeyPath = await resolveKeyPath(options.key);
    const storageDescriptor: IStorageDescriptor = {
        dbDir: databaseDir,
        encryptionKeyPath: resolvedKeyPath
    };
    
    // Get S3 config to pass to workers (needed for S3-hosted storage)
    const s3Config = await getS3Config();

    //
    // First, verify database files (metadata and sort index files) when verifying the full database.
    //
    let dbFileResult: IDatabaseFileVerifyResult | undefined;
    if (!options.path) {
        writeProgress('🗄️  Verifying database files...');
        dbFileResult = await verifyDatabaseFiles(metadataStorage, assetStorage, (progress) => {
            writeProgress(`🔍 ${progress}`);
        });
    }
    
    //
    // Then, verify asset files.
    //
    writeProgress('Verifying assets...');
    
    const result = await verify(storageDescriptor, metadataStorage, context.taskQueueProvider, metadataCollection, {
        full: options.full,
        pathFilter: options.path,
        s3Config
    }, (progress) => {
        writeProgress(`🔍 ${progress}`);
    });

    clearProgressMessage(); // Flush the progress message.

    log.info(options.path 
        ? `Verified files matching: ${options.path}` 
        : `Asset files verified.`);
    log.info('');
    
    log.info(`Files imported:   ${pc.cyan(result.totalImports.toString())}`);
    log.info(`Total files:      ${pc.cyan(result.totalFiles.toString())}`);
    log.info(`Total size:       ${pc.cyan(formatBytes(result.totalSize))}`);
    log.info(`Files processed:  ${pc.cyan(result.filesProcessed.toString())}`);
    log.info(`Nodes processed:  ${pc.cyan(result.nodesProcessed.toString())}`);
    log.info(`Unmodified:       ${pc.green(result.numUnmodified.toString())}`);
    log.info(`Modified:         ${result.modified.length > 0 ? pc.red(result.modified.length.toString()) : pc.green('0')}`);
    log.info(`New:              ${result.new.length > 0 ? pc.yellow(result.new.length.toString()) : pc.green('0')}`);
    log.info(`Removed:          ${result.removed.length > 0 ? pc.red(result.removed.length.toString()) : pc.green('0')}`);
    log.info(`Failures:         ${result.numFailures > 0 ? pc.red(result.numFailures.toString()) : pc.green('0')}`);
    log.info(`Record mismatches: ${(result.recordMismatches?.length ?? 0) > 0 ? pc.red((result.recordMismatches?.length ?? 0).toString()) : pc.green('0')}`);

    // Show details for problematic files
    if (result.modified.length > 0) {
        log.info('');
        log.info(pc.red(`Modified files:`));
        result.modified.forEach(file => {
            log.info(`  ${pc.red('●')} ${file}`);
        });
    }
    
    if (result.new.length > 0) {
        log.info('');
        log.info(pc.yellow(`New files:`));
        result.new.forEach(file => {
            log.info(`  ${pc.yellow('+')} ${file}`);
        });
    }
    
    if (result.removed.length > 0) {
        log.info('');
        log.info(pc.red(`Removed files:`));
        result.removed.forEach(file => {
            log.info(`  ${pc.red('-')} ${file}`);
        });
    }

    if ((result.recordMismatches?.length ?? 0) > 0) {
        log.info('');
        log.info(pc.red(`Asset record mismatches (missing or wrong id/hash):`));
        result.recordMismatches!.forEach(path => {
            log.info(`  ${pc.red('●')} ${path}`);
        });
    }

    log.info('');
    
    //
    // Database file summary (only when we ran database file verification, i.e. full verify without path)
    //
    if (dbFileResult !== undefined) {
        log.info(pc.bold('Database files:'));
        log.info(`  Total files:    ${pc.cyan(dbFileResult.totalFiles.toString())}`);
        log.info(`  Total size:     ${pc.cyan(formatBytes(dbFileResult.totalSize))}`);
        log.info(`  Valid files:    ${dbFileResult.validFiles === dbFileResult.totalFiles ? pc.green(dbFileResult.validFiles.toString()) : pc.yellow(dbFileResult.validFiles.toString())}`);
        log.info(`  Invalid files:  ${dbFileResult.invalidFiles.length > 0 ? pc.red(dbFileResult.invalidFiles.length.toString()) : pc.green('0')}`);
        
        // Show details for invalid database files
        if (dbFileResult.errors.length > 0) {
            log.info('');
            log.info(pc.red(`Invalid database files:`));
            for (const { file, error } of dbFileResult.errors) {
                log.info(`  ${pc.red('●')} ${file}`);
                log.info(`    ${pc.gray(error)}`);
            }
        }
        log.info('');
    }
    
    //
    // Summary
    //
    const dbFilesOk = dbFileResult === undefined || dbFileResult.invalidFiles.length === 0;
    const assetFilesOk = result.modified.length === 0 && result.new.length === 0 && result.removed.length === 0 && result.numFailures === 0 && (result.recordMismatches?.length ?? 0) === 0;
    
    if (dbFilesOk && assetFilesOk) {
        log.info(pc.green(`✅ Database verification passed - all files are intact`));
    }
    else {
        if (dbFileResult !== undefined && !dbFilesOk) {
            log.info(pc.red(`❌ Database file verification failed - ${dbFileResult.invalidFiles.length} file(s) have issues`));
        }
        if (!assetFilesOk) {
            log.info(pc.yellow(`⚠️ Asset file verification found issues - see details above`));
        }
        if ((result.recordMismatches?.length ?? 0) > 0) {
            log.info(pc.red(`❌ Asset record verification failed - ${result.recordMismatches!.length} asset(s) have missing or wrong database record`));
        }
    }

    // Show follow-up commands
    log.info('');
    log.info(pc.bold('Next steps:'));
    const hasProblems = (dbFileResult?.invalidFiles.length ?? 0) > 0 || result.modified.length > 0 || result.new.length > 0 || result.removed.length > 0 || result.numFailures > 0 || (result.recordMismatches?.length ?? 0) > 0;
    if (hasProblems) {
        log.info(pc.gray(`    # Fix database issues by restoring from source`));
        log.info(`    psi repair --source <backup-db-path>`);
        log.info('');
    }
    else {
        log.info(pc.gray(`    # Create a backup copy of your database`));
        log.info(`    psi replicate --db ${databaseDir} --dest <other-db-path>`);
        log.info('');
        log.info(pc.gray(`    # Synchronize changes between two databases that have been independently changed`));
        log.info(`    psi sync --db ${databaseDir} --dest <other-db-path>`);
        log.info('');
        log.info(pc.gray(`    # Compare this database with another location`));
        log.info(`    psi compare --db ${databaseDir} --dest <other-db-path>`);
        log.info('');
        log.info(pc.gray(`    # View database summary and tree hash`));
        log.info(`    psi summary`);
    }

    await exit(0);
}

File: apps/cli-zig/src/cmd/verify.zig
const std = @import("std");
const utils = @import("utils-zig");
const node_utils = @import("node-utils-zig");
const api = @import("api-zig");
const node_api = @import("node-api-zig");
const pc = @import("../lib/picocolors.zig");
const init_cmd = @import("../lib/init-cmd.zig");
const terminal_utils = @import("../lib/terminal-utils.zig");
const format = @import("../lib/format.zig");
const log = &utils.log.log;
const exit = node_utils.termination.exit;
const clearProgressMessage = terminal_utils.clearProgressMessage;
const writeProgress = terminal_utils.writeProgress;
const loadDatabase = init_cmd.loadDatabase;
const IBaseCommandOptions = init_cmd.IBaseCommandOptions;
const ICommandContext = init_cmd.ICommandContext;
const IDatabaseDescriptor = api.database_descriptor.IDatabaseDescriptor;
const verify = node_api.verify.verify;
const verifyDatabaseFiles = node_api.verify.verifyDatabaseFiles;
const IDatabaseFileVerifyResult = node_api.verify.IDatabaseFileVerifyResult;
const ProgressCallback = node_api.media_file_database.ProgressCallback;

//
// Options of the verify command (TypeScript: IVerifyCommandOptions extends IBaseCommandOptions).
// The base options are in `base`.
//
pub const IVerifyCommandOptions = struct {
    // The options shared by every command.
    base: IBaseCommandOptions = .{},

    //
    // Force full verification (bypass cached hash optimization).
    //
    full: ?bool = null,

    //
    // Path to a specific file or directory to verify (instead of entire database).
    //
    path: ?[]const u8 = null,
};

//
// JavaScript truthiness of an optional string.
//
fn isSet(value: ?[]const u8) bool {
    return value != null and value.?.len > 0;
}

//
// Writes each verification progress message (TypeScript: `(progress) => { writeProgress(`🔍 ${progress}`); }`).
//
fn onProgress(context: ?*anyopaque, progress: ?[]const u8) void {
    _ = context;
    var buffer: [4096]u8 = undefined;
    const message = std.fmt.bufPrint(&buffer, "\u{1F50D} {s}", .{progress orelse "undefined"}) catch return;
    writeProgress(message);
}

//
// Formats a count (`count.toString()`).
//
fn countText(allocator: std.mem.Allocator, count: u64) ![]const u8 {
    return std.fmt.allocPrint(allocator, "{d}", .{count});
}

//
// Command that verifies the integrity of the Photosphere media file database.
//
pub fn verifyCommand(allocator: std.mem.Allocator, io: std.Io, context: ICommandContext, options: *IVerifyCommandOptions) !void {
    const uuidGenerator = context.uuidGenerator;
    const timestampProvider = context.timestampProvider;
    const sessionId = context.sessionId;
    const loaded = try loadDatabase(allocator, io, options.base.db, &options.base, uuidGenerator, timestampProvider, sessionId, false);
    const assetStorage = loaded.assetStorage;
    const databaseDir = loaded.databaseDir;
    const metadataCollection = loaded.metadataCollection;

    const storageDescriptor: IDatabaseDescriptor = .{
        .databasePath = databaseDir,
        .encryptionKey = options.base.key,
    };

    const progressCallback: ProgressCallback = .{ .context = null, .function = onProgress };

    //
    // First, verify database files (metadata and sort index files) when verifying the full database.
    //
    var dbFileResult: ?IDatabaseFileVerifyResult = null;
    if (!isSet(options.path)) {
        writeProgress("\u{1F5C4}\u{FE0F}  Verifying database files...");
        dbFileResult = try verifyDatabaseFiles(allocator, io, assetStorage, progressCallback);
    }

    //
    // Then, verify asset files.
    //
    writeProgress("Verifying assets...");

    const result = try verify(allocator, io, storageDescriptor, assetStorage, context.uuidGenerator, metadataCollection, .{
        .full = options.full,
        .pathFilter = options.path,
    }, progressCallback);

    clearProgressMessage(); // Flush the progress message.

    log.info(if (isSet(options.path))
        try std.fmt.allocPrint(allocator, "Verified files matching: {s}", .{options.path.?})
    else
        "Asset files verified.");
    log.info("");

    const recordMismatchCount: u64 = if (result.recordMismatches) |recordMismatches| recordMismatches.len else 0;

    log.info(try std.fmt.allocPrint(allocator, "Files imported:    {s}", .{try pc.cyan(allocator, try countText(allocator, result.totalImports))}));
    log.info(try std.fmt.allocPrint(allocator, "Total files:       {s}", .{try pc.cyan(allocator, try countText(allocator, result.totalFiles))}));
    log.info(try std.fmt.allocPrint(allocator, "Total size:        {s}", .{try pc.cyan(allocator, try format.formatBytes(allocator, @floatFromInt(result.totalSize), format.defaultFormatBytesOptions))}));
    log.info(try std.fmt.allocPrint(allocator, "Files processed:   {s}", .{try pc.cyan(allocator, try countText(allocator, result.filesProcessed))}));
    log.info(try std.fmt.allocPrint(allocator, "Nodes processed:   {s}", .{try pc.cyan(allocator, try countText(allocator, result.nodesProcessed))}));
    log.info(try std.fmt.allocPrint(allocator, "Unmodified:        {s}", .{try pc.green(allocator, try countText(allocator, result.numUnmodified))}));
    log.info(try std.fmt.allocPrint(allocator, "Modified:          {s}", .{if (result.modified.len > 0) try pc.red(allocator, try countText(allocator, result.modified.len)) else try pc.green(allocator, "0")}));
    log.info(try std.fmt.allocPrint(allocator, "New:               {s}", .{if (result.new.len > 0) try pc.yellow(allocator, try countText(allocator, result.new.len)) else try pc.green(allocator, "0")}));
    log.info(try std.fmt.allocPrint(allocator, "Removed:           {s}", .{if (result.removed.len > 0) try pc.red(allocator, try countText(allocator, result.removed.len)) else try pc.green(allocator, "0")}));
    log.info(try std.fmt.allocPrint(allocator, "Failures:          {s}", .{if (result.numFailures > 0) try pc.red(allocator, try countText(allocator, result.numFailures)) else try pc.green(allocator, "0")}));
    log.info(try std.fmt.allocPrint(allocator, "Record mismatches: {s}", .{if (recordMismatchCount > 0) try pc.red(allocator, try countText(allocator, recordMismatchCount)) else try pc.green(allocator, "0")}));

    // Show details for problematic files
    if (result.modified.len > 0) {
        log.info("");
        log.info(try pc.red(allocator, "Modified files:"));
        for (result.modified) |file| {
            log.info(try std.fmt.allocPrint(allocator, "  {s} {s}", .{ try pc.red(allocator, "\u{25CF}"), file }));
        }
    }

    if (result.new.len > 0) {
        log.info("");
        log.info(try pc.yellow(allocator, "New files:"));
        for (result.new) |file| {
            log.info(try std.fmt.allocPrint(allocator, "  {s} {s}", .{ try pc.yellow(allocator, "+"), file }));
        }
    }

    if (result.removed.len > 0) {
        log.info("");
        log.info(try pc.red(allocator, "Removed files:"));
        for (result.removed) |file| {
            log.info(try std.fmt.allocPrint(allocator, "  {s} {s}", .{ try pc.red(allocator, "-"), file }));
        }
    }

    if (recordMismatchCount > 0) {
        log.info("");
        log.info(try pc.red(allocator, "Asset record mismatches (missing or wrong id/hash):"));
        for (result.recordMismatches.?) |path| {
            log.info(try std.fmt.allocPrint(allocator, "  {s} {s}", .{ try pc.red(allocator, "\u{25CF}"), path }));
        }
    }

    log.info("");

    //
    // Database file summary (only when we ran database file verification, i.e. full verify without path)
    //
    if (dbFileResult) |fileResult| {
        log.info(try pc.bold(allocator, "Database files:"));
        log.info(try std.fmt.allocPrint(allocator, "  Total files:    {s}", .{try pc.cyan(allocator, try countText(allocator, fileResult.totalFiles))}));
        log.info(try std.fmt.allocPrint(allocator, "  Total size:     {s}", .{try pc.cyan(allocator, try format.formatBytes(allocator, @floatFromInt(fileResult.totalSize), format.defaultFormatBytesOptions))}));
        log.info(try std.fmt.allocPrint(allocator, "  Valid files:    {s}", .{if (fileResult.validFiles == fileResult.totalFiles) try pc.green(allocator, try countText(allocator, fileResult.validFiles)) else try pc.yellow(allocator, try countText(allocator, fileResult.validFiles))}));
        log.info(try std.fmt.allocPrint(allocator, "  Invalid files:  {s}", .{if (fileResult.invalidFiles.len > 0) try pc.red(allocator, try countText(allocator, fileResult.invalidFiles.len)) else try pc.green(allocator, "0")}));

        // Show details for invalid database files
        if (fileResult.errors.len > 0) {
            log.info("");
            log.info(try pc.red(allocator, "Invalid database files:"));
            for (fileResult.errors) |fileError| {
                log.info(try std.fmt.allocPrint(allocator, "  {s} {s}", .{ try pc.red(allocator, "\u{25CF}"), fileError.file }));
                log.info(try std.fmt.allocPrint(allocator, "    {s}", .{fileError.@"error"}));
            }
        }
        log.info("");
    }

    //
    // Summary
    //
    const dbFilesOk = dbFileResult == null or dbFileResult.?.invalidFiles.len == 0;
    const assetFilesOk = result.modified.len == 0 and result.new.len == 0 and result.removed.len == 0 and result.numFailures == 0 and recordMismatchCount == 0;

    if (dbFilesOk and assetFilesOk) {
        log.info(try pc.green(allocator, "\u{2705} Database verification passed - all files are intact"));
    }
    else {
        if (dbFileResult != null and !dbFilesOk) {
            log.info(try pc.red(allocator, try std.fmt.allocPrint(allocator, "\u{274C} Database file verification failed - {d} file(s) have issues", .{dbFileResult.?.invalidFiles.len})));
        }
        if (!assetFilesOk) {
            log.info(try pc.yellow(allocator, "\u{26A0}\u{FE0F} Asset file verification found issues - see details above"));
        }
        if (recordMismatchCount > 0) {
            log.info(try pc.red(allocator, try std.fmt.allocPrint(allocator, "\u{274C} Asset record verification failed - {d} asset(s) have missing or wrong database record", .{recordMismatchCount})));
        }
    }

    // Show follow-up commands
    log.info("");
    log.info(try pc.bold(allocator, "Next steps:"));
    const invalidDbFileCount: u64 = if (dbFileResult) |fileResult| fileResult.invalidFiles.len else 0;
    const hasProblems = invalidDbFileCount > 0 or result.modified.len > 0 or result.new.len > 0 or result.removed.len > 0 or result.numFailures > 0 or recordMismatchCount > 0;
    if (hasProblems) {
        log.info("    # Fix database issues by restoring from source");
        log.info("    psi repair --source <backup-db-path>");
        log.info("");
    }
    else {
        log.info("    # Create a backup copy of your database");
        log.info(try std.fmt.allocPrint(allocator, "    psi replicate --db {s} --dest <other-db-path>", .{databaseDir}));
        log.info("");
        log.info("    # Synchronize changes between two databases that have been independently changed");
        log.info(try std.fmt.allocPrint(allocator, "    psi sync --db {s} --dest <other-db-path>", .{databaseDir}));
        log.info("");
        log.info("    # Compare this database with another location");
        log.info(try std.fmt.allocPrint(allocator, "    psi compare --db {s} --dest <other-db-path>", .{databaseDir}));
        log.info("");
        log.info("    # View database summary and tree hash");
        log.info("    psi summary");
    }

    exit(io, 0);
}
